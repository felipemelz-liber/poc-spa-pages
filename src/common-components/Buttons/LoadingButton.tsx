import React, { useRef, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { SuccessButton, LoadingSpinner } from 'liber-components';
import { LoadingContainer } from './LoadingButton.styles';

const LoadingButton = ({
  ButtonComponent,
  children,
  loading,
  onClick,
  disabled,
  size,
  spinnerColor,
  ...rest
}) => {
  const buttonRef = useRef(null);
  const [buttonWidth, setButtonWidth] = useState(0);

  useEffect(() => {
    setButtonWidth(buttonRef.current ? buttonRef.current.offsetWidth : 0);
  }, [buttonRef.current]);

  const getSpinnerSize = () => (size === 'small' ? '14px' : '22px');
  return (
    <ButtonComponent
      ref={buttonRef}
      size={size}
      version={2}
      onClick={onClick}
      disabled={disabled || loading}
      {...rest}
    >
      {loading ? (
        <LoadingContainer minWidth={buttonWidth - 32}>
          <LoadingSpinner color={spinnerColor} size={getSpinnerSize()} />
        </LoadingContainer>
      ) : (
        children
      )}
    </ButtonComponent>
  );
};

LoadingButton.propTypes = {
  ButtonComponent: PropTypes.elementType,
  children: PropTypes.node,
  loading: PropTypes.bool,
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  spinnerColor: PropTypes.string,
};

LoadingButton.defaultProps = {
  ButtonComponent: SuccessButton,
  children: 'BotÃ£o',
  loading: false,
  onClick: () => null,
  disabled: false,
  size: 'medium',
  spinnerColor: 'white',
};

export default LoadingButton;
