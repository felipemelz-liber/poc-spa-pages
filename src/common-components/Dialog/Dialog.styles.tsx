import styled from 'styled-components';

export const DialogHeader = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 24px;
`;

export const DialogActions = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-end;
  padding: 16px 24px;
  padding-top: 0px;
`;

export const DialogContent = styled.div`
  padding: 24px;
  padding-top: 0px;
  max-width: ${({ maxWidth }) => maxWidth || '522'}px;
`;

export const Title = styled.span`
  font-family: Roboto;
  font-size: 21px;
  font-weight: 500;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.4;
  letter-spacing: 0.08px;
  text-align: left;
  color: #4b6f85;
`;
