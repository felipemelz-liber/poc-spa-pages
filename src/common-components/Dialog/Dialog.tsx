import React from 'react';
import PropTypes from 'prop-types';
import { Modal, FlatButton } from 'liber-components';
import { DialogHeader, Title, DialogContent, DialogActions } from './Dialog.styles';
import LoadingButton from '../Buttons/LoadingButton';

const Dialog = ({ open, onClose, confirmationButtonProps, title, children, cancelLabel }) => {
  const { label, onConfirm, color, disabled, loading } = confirmationButtonProps;
  return (
    <Modal show={open} onLeaved={onClose}>
      <DialogHeader>
        <Title>{title}</Title>
      </DialogHeader>
      <DialogContent>{children}</DialogContent>
      <DialogActions>
        <FlatButton onClick={onClose}>{cancelLabel}</FlatButton>
        <LoadingButton
          ButtonComponent={FlatButton}
          color={color}
          onClick={onConfirm}
          disabled={disabled}
          loading={loading}
          spinnerColor={color === 'primary' ? '#009dff' : '#da2d16'}
          size="small"
        >
          {label || 'CONTINUAR'}
        </LoadingButton>
      </DialogActions>
    </Modal>
  );
};

Dialog.propTypes = {
  open: PropTypes.bool,
  onClose: PropTypes.func,
  confirmationButtonProps: PropTypes.shape({
    onConfirm: PropTypes.func,
    label: PropTypes.string,
    color: PropTypes.oneOf(['primary', 'danger']),
    disabled: PropTypes.bool,
    loading: PropTypes.bool,
  }),
  title: PropTypes.string,
  children: PropTypes.node,
  cancelLabel: PropTypes.string,
};

Dialog.defaultProps = {
  open: false,
  onClose: () => {},
  confirmationButtonProps: {
    onConfirm: () => {},
    label: 'CONTINUAR',
    color: 'primary',
    disabled: false,
    loading: false,
  },
  title: '',
  children: null,
  cancelLabel: 'CANCELAR',
};

export default Dialog;
