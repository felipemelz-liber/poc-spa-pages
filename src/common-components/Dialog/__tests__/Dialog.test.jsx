import React from 'react';
import sinon from 'sinon';
import { shallow, configure } from 'enzyme';
import { Modal, FlatButton } from 'liber-components';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import Dialog from '../Dialog';
import LoadingButton from '../../Buttons/LoadingButton';

describe('Dialog tests', () => {
  configure({ adapter: new Adapter() });
  const onClose = sinon.spy();
  const onConfirm = sinon.spy();

  const props = {
    onClose,
    confirmationButtonProps: {
      onConfirm,
    },
  };

  afterEach(() => {
    sinon.resetHistory();
  });

  it('should render correctly without props', () => {
    const wrapper = shallow(<Dialog />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call onClose on modal leave', () => {
    const wrapper = shallow(<Dialog {...props} />);
    wrapper.find(Modal).simulate('leaved');
    expect(onClose.called).toBeTruthy();
  });

  it('should call onClose on CancelButton click', () => {
    const wrapper = shallow(<Dialog {...props} />);
    wrapper.find(FlatButton).simulate('click');
    expect(onClose.called).toBeTruthy();
  });

  it('should call onConfirm on ConfirmButton click', () => {
    const wrapper = shallow(<Dialog {...props} />);
    wrapper.find(LoadingButton).simulate('click');
    expect(onConfirm.called).toBeTruthy();
  });

  it('should disable ConfirmButton if disabled is true on confirmationButtonProps', () => {
    const disabledProps = {
      confirmationButtonProps: {
        disabled: true,
      },
    };
    const wrapper = shallow(<Dialog {...disabledProps} />);
    const confirmButton = wrapper.find(LoadingButton);
    expect(confirmButton.props().disabled).toBeTruthy();
  });
});
