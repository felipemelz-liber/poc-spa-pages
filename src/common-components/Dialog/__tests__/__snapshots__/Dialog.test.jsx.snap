// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog tests should render correctly without props 1`] = `
<Modal
  animateFrom="-150px"
  animation={
    Object {
      "in": "cubic-bezier(.17,1.92,.3,.71)",
      "out": "cubic-bezier(1,.23,.76,-1.11)",
    }
  }
  closable={true}
  closeButton={false}
  closeButtonPadding="10px 14px"
  duration={270}
  elementId="modal"
  fullscreen={false}
  insertComponent={null}
  label=""
  labelAlign="left"
  onAppearStart={[Function]}
  onAppeared={[Function]}
  onLeaveStart={[Function]}
  onLeaved={[Function]}
  overflow=""
  position={null}
  show={false}
>
  <styled.div>
    <styled.span />
  </styled.div>
  <styled.div />
  <styled.div>
    <StyledButtons__FlatButton
      color="primary"
      onClick={[Function]}
    >
      CANCELAR
    </StyledButtons__FlatButton>
    <LoadingButton
      ButtonComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "_foldedDefaultProps": Object {
            "color": "primary",
          },
          "attrs": Array [
            [Function],
          ],
          "componentStyle": ComponentStyle {
            "componentId": "StyledButtons__FlatButton-sc-1e7va4d-2",
            "isStatic": false,
            "rules": Array [
              "--inner-button-text-color:var(--button-text-color,",
              [Function],
              ");--inner-button-ripple-color:var(--button-ripple-color,",
              [Function],
              ");padding:24px 0px;",
              [Function],
              " ",
              [Function],
              " background-color:",
              [Function],
              ";color:var(--inner-button-text-color);border:",
              [Function],
              ";border-radius:",
              [Function],
              ";font-family:",
              [Function],
              ";cursor:pointer;padding:0 16px;transition:all 450ms cubic-bezier(0.23,1,0.32,1) 0ms;box-sizing:border-box;display:inline-flex;align-items:center;justify-content:center;box-sizing:border-box;position:relative;overflow:hidden;&:focus{outline:0;}&::-moz-focus-inner{border:0;}&:disabled{opacity:0.38;cursor:default;pointer-events:none;}&:after{content:'';--ripple-color:;position:absolute;top:50%;left:50%;width:5px;height:5px;background:var(--inner-button-ripple-color);opacity:0;border-radius:100%;transform:scale(1,1) translate(-50%);transform-origin:50% 50%;}&:focus:not(:active)::after{animation:",
              Keyframes {
                "id": "sc-keyframes-kHeZab",
                "inject": [Function],
                "name": "kHeZab",
                "rules": Array [
                  "@-webkit-keyframes kHeZab{0%{-webkit-transform:scale(0,0);-ms-transform:scale(0,0);transform:scale(0,0);opacity:0.8;}20%{-webkit-transform:scale(25,25);-ms-transform:scale(25,25);transform:scale(25,25);opacity:0.8;}100%{opacity:0;-webkit-transform:scale(40,40);-ms-transform:scale(40,40);transform:scale(40,40);}}",
                  "@keyframes kHeZab{0%{-webkit-transform:scale(0,0);-ms-transform:scale(0,0);transform:scale(0,0);opacity:0.8;}20%{-webkit-transform:scale(25,25);-ms-transform:scale(25,25);transform:scale(25,25);opacity:0.8;}100%{opacity:0;-webkit-transform:scale(40,40);-ms-transform:scale(40,40);transform:scale(40,40);}}",
                ],
                "toString": [Function],
              },
              " 0.8s ease-out;}& > svg{min-width:var(--icon-size,16px)!important;max-width:var(--icon-size,16px)!important;min-height:var(--icon-size,16px)!important;max-height:var(--icon-size,16px)!important;fill:var(--inner-button-text-color);margin-right:var(--svg-margin,12px);}",
              "background-color:transparent;",
              [Function],
            ],
          },
          "displayName": "StyledButtons__FlatButton",
          "foldedComponentIds": Array [
            "StyledButtons__Button-sc-1e7va4d-0",
          ],
          "propTypes": Object {
            "color": [Function],
          },
          "render": [Function],
          "styledComponentId": "StyledButtons__FlatButton-sc-1e7va4d-2",
          "target": "button",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      color="primary"
      disabled={false}
      loading={false}
      onClick={[Function]}
      size="medium"
      spinnerColor="#009dff"
    >
      CONTINUAR
    </LoadingButton>
  </styled.div>
</Modal>
`;
